### PR TITLE
Add version file compatible with other Key4hep packages

### PR DESCRIPTION
**Pull request by tmadlener**
BEGINRELEASENOTES
- Add a `EDM4hepVersion.h` file that has the same basic structure and functionality as other Key4hep packages.

ENDRELEASENOTES

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -89,9 +89,9 @@ endif()
 
 #--- add version files ---------------------------------------------------------
 
-configure_file(${CMAKE_CURRENT_SOURCE_DIR}/EDM4HEPVersion.h
-               ${CMAKE_CURRENT_BINARY_DIR}/EDM4HEPVersion.h )
-install(FILES ${CMAKE_CURRENT_BINARY_DIR}/EDM4HEPVersion.h
+configure_file(${CMAKE_CURRENT_SOURCE_DIR}/EDM4hepVersion.h.in
+               ${CMAKE_CURRENT_BINARY_DIR}/EDM4hepVersion.h )
```
Comment (tmadlener) [inline CMakeLists.txt:93]:
```suggestion
               ${CMAKE_CURRENT_BINARY_DIR}/edm4hep/EDM4hepVersion.h )
```
Maybe this should go into `edm4hep` as well? In order to be compatible with where we install it to.

**Comment by vvolkl** [comment]
@tmadlener Is there a way to write this version to the metadata  as well? Would probably need to be done in podio or in the user code?

File: CMakeLists.txt (hunk 1 above)
Comment (vvolkl) [reply CMakeLists.txt:93]: Makes sense!

**Comment by tmadlener** [comment]
For podio we are already writing the build version of podio. However, there is no "easy" way yet to do it for a generated datamodel like EDM4hep. What could be done at the moment is to write the 64 bit encoded `unsigned` as user data, or potentially also writing it as 3 `int`s into the meta data. None of these are ideal, I think and this should be done automatically without having to do it explicitly.

Since writing the version is necessary in any case for schema evolution, this has to be solved in any case rather soon. With that in mind, maybe the current header has to be considered as not completely stable yet.

**Review by vvolkl — APPROVED** [review]

**Comment by tmadlener** [comment]
Merging this as it is at the moment, and decide how to deal with the schema version once that is actually available.

File: EDM4hepVersion.h.in [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+#include <ostream>
+#if __cplusplus >= 202002L
+#include <compare>
+#endif
+
+// Some preprocessor constants and macros for the use cases where they might be
+// necessary
+
+/// Define a version to be used in edm4hep.
+#define EDM4HEP_VERSION(major, minor, patch) (((unsigned long)(major) << 32) | ((unsigned long)(minor) << 16) | ((unsigned long)(patch)))
```
Comment (tmadlener) [inline EDM4hepVersion.h.in:15]: This does not work when used in the preprocessor. See AIDASoft/podio#374

Comment (tmadlener) [inline EDM4hepVersion.h.in:15]:
```suggestion
#define EDM4HEP_VERSION(major, minor, patch) ((UINT64_C(major) << 32) | (UINT64_C(minor) << 16) | (UINT64_C(patch)))
```

File: EDM4hepVersion.h.in [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+/// Get the major version from a preprocessor defined version
+#define EDM4HEP_MAJOR_VERSION(v) (((v) & (-1UL >> 16)) >> 32)
+/// Get the minor version from a preprocessor defined version
+#define EDM4HEP_MINOR_VERSION(v) (((v) & (-1UL >> 32)) >> 16)
+/// Get the patch version from a preprocessor defined version
+#define EDM4HEP_PATCH_VERSION(v) ((v) & (-1UL >> 48))
+
+// Some helper constants that are populated by the cmake configure step
+#define edm4hep_VERSION @edm4hep_VERSION@
```
Comment (tmadlener) [inline EDM4hepVersion.h.in:24]:
```suggestion
```

File: EDM4hepVersion.h.in [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+#define EDM4HEP_VERSION_MAJOR @EDM4HEP_VERSION_MAJOR@
+#define EDM4HEP_VERSION_MINOR @EDM4HEP_VERSION_MINOR@
+#define EDM4HEP_VERSION_PATCH @EDM4HEP_VERSION_PATCH@
+
```
Comment (tmadlener) [inline EDM4hepVersion.h.in:28]:
```suggestion
#define edm4hep_VERSION EDM4HEP_VERSION(EDM4HEP_VERSION_MAJOR, EDM4HEP_VERSION_MINOR, EDM4HEP_VERSION_PATCH)
```
Also make this more useful than what it was before.

File: EDM4hepVersion.h.in [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  /**
+   * The current build version
+   */
+  static constexpr Version build_version{EDM4HEP_VERSION_MAJOR, EDM4HEP_VERSION_MINOR, EDM4HEP_VERSION_PATCH};
+
+  /**
+   * Decode a version from a 64 bit unsigned
+   */
+  static constexpr Version decode_version(unsigned long version) noexcept {
```
Comment (tmadlener) [inline EDM4hepVersion.h.in:78]:
```suggestion
  static constexpr Version decode_version(uint64_t version) noexcept {
```

**Comment by tmadlener** [comment]
Meeting: Podio could generate this file (and add schema information into it)

**Comment by hegner** [comment]
I think that would be a slightly bigger thing to discuss. feature creep is a big risk there...

**Pull request merged**